{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "534056fb-f5d6-4f7a-ae0d-ab8b49d3c430",
   "metadata": {},
   "source": [
    "1. load the function\n",
    "2. read the data (vcfs, parsed.flow, parsed.geno, args)\n",
    "3. make plots\n",
    "4. save"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c9b4ed2-470e-407f-a70f-7d210de3233c",
   "metadata": {},
   "source": [
    "# Load functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "cab09e3a-c0f1-43eb-b831-28384b063834",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library('vcfR'))\n",
    "suppressMessages(library('data.table'))\n",
    "suppressMessages(library('RColorBrewer'))\n",
    "suppressMessages(library('kinship2'))\n",
    "suppressMessages(library('plotly'))\n",
    "suppressMessages(library('htmltools'))\n",
    "suppressMessages(library('GenomicRanges'))\n",
    "suppressMessages(library('DNAcopy'))\n",
    "suppressMessages(library('knitr'))\n",
    "suppressMessages(library('stringr'))\n",
    "suppressMessages(library(htmlwidgets))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "f637bcc2-6e24-4010-b962-02e69d58b134",
   "metadata": {},
   "outputs": [],
   "source": [
    "args <- list(\n",
    "  ## mandatory arguments\n",
    "  vcf.file=c(),\n",
    "  sample.ids=c(),\n",
    "  \n",
    "  ## important optional arguments\n",
    "  father.ids=c(),\n",
    "  mother.ids=c(),\n",
    "  genders=c(),\n",
    "  run.merlin=T,\n",
    "  cytoband.file=c(),\n",
    "  \n",
    "  ## variant inclusion arguments: filter 1\n",
    "  dp.hard.limit.ids=c(),\n",
    "  dp.hard.limit=10,\n",
    "  af.hard.limit.ids=c(),\n",
    "  af.hard.limit=0,\n",
    "  dp.soft.limit.ids=c(),\n",
    "  dp.soft.limit=10,\n",
    "  \n",
    "  ## variant inclusion arguments: filter 2\n",
    "  keep.informative.ids=c(),\n",
    "  keep.hetero.ids=c(),\n",
    "  \n",
    "  ## sample/disease annotation\n",
    "  regions=c(),\n",
    "  reference.ids=c(),\n",
    "  carrier.ids=c(),\n",
    "  affected.ids=c(),\n",
    "  nonaffected.ids=c(),\n",
    "  info=c(),\n",
    "  \n",
    "  ## BAF profiles\n",
    "  baf.ids=c(),\n",
    "  \n",
    "  ## merlin profiles\n",
    "  merlin.model='best',\n",
    "  min.seg.var=5,\n",
    "  min.seg.var.X=15,\n",
    "  window.size.voting=10000000,\n",
    "  window.size.voting.X=c(),\n",
    "  keep.chromosomes.only=T,\n",
    "  keep.regions.only=F,\n",
    "  concordance.table=T,\n",
    "  \n",
    "  ## remaining features\n",
    "  out.dir=c('./'),\n",
    "  fam.id='hopla',\n",
    "  X.cutoff=1.5,\n",
    "  Y.cutoff=.6,\n",
    "  window.size=1000000,\n",
    "  regions.flanking.size=2000000,\n",
    "  limit.baf.to.P=F,\n",
    "  limit.pm.to.P=F,\n",
    "  value.of.P=.25,\n",
    "  color.palette='Paired',\n",
    "  dot.factor=2,\n",
    "  self.contained=F,\n",
    "  cairo=F\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e0bae67f-f33b-4bd5-93c4-a1c9f944dfd4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions for loading arguments\n",
    "\n",
    "trim <- function (x) gsub(\"^\\\\s+|\\\\s+$\", \"\", x)\n",
    "\n",
    "format.value <- function(arg, value){\n",
    "  if (!(arg %in% names(args))){\n",
    "    cat(paste0('ERROR: Argument --', arg, ' does not exist.\\n'))\n",
    "    quit(status=0)\n",
    "  }\n",
    "  \n",
    "  numeric.args <- c('dp.hard.limit', 'af.hard.limit', 'dp.soft.limit',\n",
    "                    'min.seg.var', 'min.seg.var.X', 'window.size.voting',\n",
    "                    'window.size.voting.X', 'X.cutoff', 'Y.cutoff', 'window.size',\n",
    "                    'regions.flanking.size', 'value.of.P', 'dot.factor')\n",
    "  boolean.args <- c('run.merlin', 'keep.chromosomes.only', 'keep.regions.only',\n",
    "                    'concordance.table', 'limit.baf.to.P', 'limit.pm.to.P',\n",
    "                    'self.contained', 'cairo')\n",
    "  \n",
    "  value = sapply(strsplit(value, ',')[[1]], function(x) trim(x))\n",
    "  \n",
    "  ## replace NAs\n",
    "  value[value %in% c('', 'NA')] = NA\n",
    "  \n",
    "  ## tranform when needed\n",
    "  if (arg %in% numeric.args) value = as.numeric(value)\n",
    "  if (arg %in% boolean.args) value = as.logical(value)\n",
    "  return(value)\n",
    "}\n",
    "\n",
    "get.file.args <- function(settings.file){\n",
    "  if (!file.exists(settings.file)){\n",
    "    cat('ERROR: File given by --settings does not exist. Please Correct.\\n')\n",
    "    quit(status=0)\n",
    "  }\n",
    "  at.info = F\n",
    "  for (line in suppressWarnings(readLines(settings.file))){\n",
    "    ## info parsing\n",
    "    if (!at.info) line = gsub(\"'\", '', gsub('\"', '', line))\n",
    "    if (at.info) line = gsub('\\t', '    ', line)\n",
    "    if (line == 'end.info'){ at.info = F ; next }\n",
    "    if (at.info){ args$info <- c(args$info, line) ; next }\n",
    "    if (line == 'start.info'){ at.info = T ; next }\n",
    "    \n",
    "    ## skip comments\n",
    "    if (substr(trim(line), 1, 1) %in% c('#', '')) next\n",
    "    line <- trim(strsplit(line, '#', fixed = T)[[1]][1])\n",
    "    \n",
    "    ## parse argument\n",
    "    arg <- trim(strsplit(line, '=')[[1]][1])\n",
    "    value <- strsplit(line, '=')[[1]][2]\n",
    "    value <- format.value(arg, value)\n",
    "    \n",
    "    if (substr(line, nchar(line), nchar(line)) != '=' & \n",
    "        grepl('=', line, fixed = T)) args[[arg]] <- value\n",
    "  }\n",
    "  return(args)\n",
    "}\n",
    "\n",
    "post.process.args <- function(args){\n",
    "  no.u.mask <- !sapply(args$sample.ids, function(x) toupper(substr(x,1,1)) == 'U' &\n",
    "                         !is.na(suppressWarnings(as.numeric(substr(x,2,999)))))\n",
    "  \n",
    "  if (!length(args$genders)) args$genders = rep(NA, length(args$sample.ids))\n",
    "  if (!length(args$mother.ids)) args$mother.ids = rep(NA, length(args$sample.ids))\n",
    "  if (!length(args$father.ids)) args$father.ids = rep(NA, length(args$sample.ids))\n",
    "  \n",
    "  args$samples.no.u <- args$sample.ids[no.u.mask]\n",
    "  args$samples.u <- args$sample.ids[!no.u.mask]\n",
    "  \n",
    "  not.last.in.line <- unique(c(args$mother.ids, args$father.ids))\n",
    "  not.last.in.line <- not.last.in.line[not.last.in.line %in% args$samples.no.u]\n",
    "  last.in.line <- args$samples.no.u[!args$samples.no.u %in% not.last.in.line]\n",
    "  \n",
    "  if (length(args$samples.no.u) == 1){\n",
    "    not.last.in.line = args$samples.no.u\n",
    "    last.in.line = args$samples.no.u\n",
    "  }\n",
    "  \n",
    "  if (!length(args$dp.hard.limit.ids)) args$dp.hard.limit.ids = not.last.in.line\n",
    "  if (!length(args$af.hard.limit.ids)) args$af.hard.limit.ids = not.last.in.line\n",
    "  if (!length(args$dp.soft.limit.ids)) args$dp.soft.limit.ids = last.in.line\n",
    "  if (!length(args$window.size.voting.X)) args$window.size.voting.X = args$window.size.voting\n",
    "  if (!length(args$min.seg.var.X)) args$min.seg.var.X = args$min.seg.var\n",
    "  \n",
    "  man.args <- c('vcf.file', 'sample.ids')\n",
    "  for (arg in names(args)){\n",
    "    if (!length(args[[arg]]) & arg %in% man.args){\n",
    "      cat(paste0('ERROR: Argument --', arg, ' is mandatory. Please provide.\\n'))\n",
    "      quit(status=0)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  not.in.error <- function(arg){\n",
    "    not.in <- function(xs, ys){\n",
    "      for (x in xs){\n",
    "        if (!(x %in% ys) & !is.na(x)){\n",
    "          return(x)\n",
    "        }\n",
    "      }\n",
    "      return(NULL)\n",
    "    }\n",
    "    if (length(not.in(args[[arg]], args$sample.ids))){\n",
    "      cat(paste0('ERROR: Fetched from argument --', arg,', \\'', not.in(args[[arg]], args$sample.ids),\n",
    "                 '\\' could not be found in the provided --sample.ids. Please correct.\\n'))\n",
    "      quit(status=0)\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  cat('Selected parameters ...\\n')\n",
    "  for (arg in names(args)[!(names(args) %in% c('samples.u', 'samples.no.u'))]){\n",
    "    if (!length(args[[arg]])) next\n",
    "    cat(paste0('  ... ', arg, ': ', paste(args[[arg]], collapse = ','), '\\n'))\n",
    "    \n",
    "    if (any(is.na(args[[arg]])) & !(arg %in% c('father.ids', 'mother.ids', 'genders'))){\n",
    "      cat(paste0('ERROR: No \\'NA\\' allowed in argument --', arg,'. Please correct.\\n'))\n",
    "      quit(status=0)\n",
    "    }\n",
    "    \n",
    "    if (arg == 'sample.ids'){\n",
    "      if (length(args$sample.ids) > 1){\n",
    "        if (!length(which(!is.na(args$mother.ids))) & !length(which(!is.na(args$father.ids)))){\n",
    "          cat('ERROR: More than one sample is given in --sample.ids. Please provide their relation using argument(s)', \n",
    "              '--father.ids and/or --mother.ids. Otherwise, run separately.\\n')\n",
    "          quit(status=0)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if (arg == 'genders'){\n",
    "      if (!all(args$genders %in% c('M', 'F', NA))){\n",
    "        cat('ERROR: Argument --genders should be coded as \\'M\\', \\'F\\' or \\'NA\\'. Please correct.\\n')\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    for (same.length.arg in c('father.ids', 'mother.ids', 'genders')){\n",
    "      if (arg == same.length.arg){\n",
    "        if (!(length(args$sample.ids) == length(args[[same.length.arg]]))){\n",
    "          cat(paste0('ERROR: Arguments --sample.ids and --', same.length.arg,' should be of the same length. Please correct.\\n'))\n",
    "          quit(status=0)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if (arg %in% c('father.ids', 'mother.ids', 'dp.hard.limit.ids', 'af.hard.limit.ids',\n",
    "                  'dp.soft.limit.ids', 'keep.informative.ids', 'keep.hetero.ids',\n",
    "                  'reference.ids', 'carrier.ids', 'affected.ids',\n",
    "                  'nonaffected.ids', 'baf.ids')){\n",
    "      not.in.error(arg)\n",
    "    }\n",
    "    \n",
    "    if (arg %in% c('dp.hard.limit.ids', 'af.hard.limit.ids', 'dp.soft.limit.ids',\n",
    "                   'keep.informative.ids', 'keep.hetero.ids', 'baf.ids')){\n",
    "      if (length(intersect(args[[arg]], args$samples.u))){\n",
    "        cat(paste0('ERROR: \\'U\\' IDs not allowed in --', arg, '. Please correct.\\n'))\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if (arg == 'run.merlin' & args$run.merlin){\n",
    "      if (length(args$sample.ids) == 1){\n",
    "        cat(paste0('WARNING: Only one sample provided. Setting --run.merlin FALSE.\\n'))\n",
    "        args$run.merlin = F\n",
    "      }\n",
    "      if ((Sys.which('merlin') == '' | Sys.which('minx') == '') & args$run.merlin){\n",
    "        cat(paste0('WARNING: Merlin executables folder could not be located in $PATH. Setting --run.merlin FALSE.\\n'))\n",
    "        args$run.merlin = F\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if (arg == 'keep.informative.ids'){\n",
    "      if (!(length(args[[arg]]) %in% c(0,2))){\n",
    "        cat(paste0('ERROR: No or two samples should be given at --keep.informative.ids. Please correct.\\n'))\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if (arg == 'merlin.model'){\n",
    "      if (!(args$merlin.model %in% c('sample', 'best'))){\n",
    "        cat('ERROR: Argument --merlin.model should be coded as \\'sample\\' or \\'best\\'. Please correct.\\n')\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    if (arg == 'vcf.file'){\n",
    "      if (!file.exists(args$vcf.file)){\n",
    "        cat('ERROR: the file given by --vcf.file does not exist. Please correct.\\n')\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    if (arg == 'cytoband.file'){\n",
    "      if (length(args$cytoband.file) & !file.exists(args$cytoband.file)){\n",
    "        cat('ERROR: the file given by --vcf.file does not exist. Please correct.\\n')\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    if (arg == 'value.of.P'){\n",
    "      if (args$value.of.P <= 0 | args$value.of.P > 1){\n",
    "        cat('ERROR: --value.of.P should be within ]0, 1]. Please correct.\\n')\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "    if (arg == 'af.hard.limit'){\n",
    "      if (args$af.hard.limit < 0 | args$af.hard.limit >= 1){\n",
    "        cat('ERROR: --af.hard.limit should be within [0, 1[. Please correct.\\n')\n",
    "        quit(status=0)\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  add.annot <- function(letter, annot = NULL){\n",
    "    if (!is.null(annot)){\n",
    "      args$samples.out[args$sample.ids %in% args[[annot]]] <- paste0(args$samples.out[args$sample.ids %in% args[[annot]]],\n",
    "                                                                     ' (', letter, ')')\n",
    "    } else {\n",
    "      known.set <- unique(c(args$reference.ids, args$carrier.ids, args$affected.ids, args$nonaffected.ids))\n",
    "      args$samples.out[!(args$sample.ids %in% known.set)] <- paste0(args$samples.out[!(args$sample.ids %in% known.set)],\n",
    "                                                                    ' (', letter, ')')\n",
    "    }\n",
    "    return(args)\n",
    "  }\n",
    "  args$samples.out <- args$sample.ids\n",
    "  args <- add.annot('R', 'reference.ids')\n",
    "  args <- add.annot('C', 'carrier.ids')\n",
    "  args <- add.annot('A', 'affected.ids')\n",
    "  args <- add.annot('N', 'nonaffected.ids')\n",
    "  \n",
    "  return(args)\n",
    "}\n",
    "\n",
    "correct.profiles <- function(args, parsed.flow){\n",
    "  is.corrected <- list()\n",
    "  for (chr in chrs){\n",
    "    is.corrected[[chr]] <- matrix(nrow = nrow(map.list[[chr]]), ncol = length(args$samples.no.u) * 2)\n",
    "    for (i in 1:length(args$samples.no.u)){\n",
    "      is.corrected[[chr]][,(i*2)-1] <- F\n",
    "      is.corrected[[chr]][,i*2] <- F\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  correct.vector.1 <- function(v, pos, max.distance){\n",
    "    letters = unique(v)\n",
    "    neighbours.i <- lapply(pos, function(p) c(which(abs(pos - p) <= max.distance)))\n",
    "    neighbours.letters <- lapply(1:length(v), function(i) v[neighbours.i[[i]]])\n",
    "    neighbours.pos <- lapply(1:length(v), function(i) abs(pos - pos[i])[neighbours.i[[i]]])\n",
    "    neighbours.weights <- lapply(neighbours.pos, function(x) (max.distance * 2) / (x + max.distance) - 1)\n",
    "    if (max.distance == 0) neighbours.weights <- lapply(1:length(v), function(x) 1)\n",
    "    c.v <- sapply(1:length(v), function(i)\n",
    "      letters[which.max(c(sum(neighbours.weights[[i]][neighbours.letters[[i]] == letters[1]]),\n",
    "                          sum(neighbours.weights[[i]][neighbours.letters[[i]] == letters[2]], na.rm = T)))])\n",
    "    return(c.v)\n",
    "  }\n",
    "  \n",
    "  correct.vector.2 <- function(flow, geno, min.seg.var){\n",
    "    breakpoints <- which(c('ZZ', flow) != c(flow, 'ZZ'))\n",
    "    for (i in 1:(length(breakpoints)-1)){\n",
    "      sequence <- breakpoints[i]:c(breakpoints[i+1]-1)\n",
    "      if (length(which(geno[sequence] != 'NA')) > min.seg.var) next\n",
    "      if (breakpoints[i]-1 != 0){\n",
    "        letter = flow[breakpoints[i]-1] # previous segment when possible\n",
    "      } else {\n",
    "        letter = flow[breakpoints[i+1]] # next segment otherwise\n",
    "      }\n",
    "      flow[sequence] <- rep(letter, length(sequence))\n",
    "    }\n",
    "    return(flow)\n",
    "  }\n",
    "  \n",
    "  if (args$window.size.voting != 0 | args$min.seg.var != 0){\n",
    "    cat('Correcting haplotypes, working ...\\n')\n",
    "    for (chr in chrs[1:22]){\n",
    "      cat(paste0('  ... at ', chr, '\\n'))\n",
    "      pos = map.list[[chr]][,2]\n",
    "      for (i in 1:length(args$samples.no.u)){\n",
    "        v = parsed.flow[[chr]][,i]\n",
    "        a <- sapply(strsplit(v, '|', fixed = T), function(x) x[1])\n",
    "        b <- sapply(strsplit(v, '|', fixed = T), function(x) x[2])\n",
    "        c.a = a\n",
    "        c.b = b\n",
    "        if (args$min.seg.var != 0){\n",
    "          c.a <- correct.vector.2(c.a, sapply(strsplit(parsed.geno[[chr]][,i], '|', fixed = T), function(x) x[1]),\n",
    "                                  args$min.seg.var)\n",
    "          c.b <- correct.vector.2(c.b, sapply(strsplit(parsed.geno[[chr]][,i], '|', fixed = T), function(x) x[2]),\n",
    "                                  args$min.seg.var)\n",
    "        }\n",
    "        if (args$window.size.voting != 0){\n",
    "          c.a <- correct.vector.1(c.a, pos, args$window.size.voting / 2)\n",
    "          c.b <- correct.vector.1(c.b, pos, args$window.size.voting / 2)\n",
    "        }\n",
    "        parsed.flow[[chr]][,i] <- paste0(c.a, '|', c.b)\n",
    "        is.corrected[[chr]][,(i*2)-1] <- a != c.a\n",
    "        is.corrected[[chr]][,i*2] <- b != c.b\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  if (args$window.size.voting.X != 0 | args$min.seg.var.X != 0){\n",
    "    if (args$window.size.voting == 0 & args$min.seg.var == 0) cat('Correcting haplotypes, working ...\\n')\n",
    "    chr = chrs[23]\n",
    "    cat(paste0('  ... at ', chr, '\\n'))\n",
    "    pos = map.list[[chr]][,2]\n",
    "    for (i in 1:length(args$samples.no.u)){\n",
    "      v = parsed.flow[[chr]][,i]\n",
    "      a <- sapply(strsplit(v, '|', fixed = T), function(x) x[1])\n",
    "      b <- sapply(strsplit(v, '|', fixed = T), function(x) x[2])\n",
    "      c.a = a\n",
    "      c.b = b\n",
    "      if (args$min.seg.var.X != 0){\n",
    "        c.a <- correct.vector.2(c.a, sapply(strsplit(parsed.geno[[chr]][,i], '|', fixed = T), function(x) x[1]),\n",
    "                                args$min.seg.var.X)\n",
    "        c.b <- correct.vector.2(c.b, sapply(strsplit(parsed.geno[[chr]][,i], '|', fixed = T), function(x) x[2]),\n",
    "                                args$min.seg.var.X)\n",
    "      }\n",
    "      if (args$window.size.voting.X != 0){\n",
    "        c.a <- correct.vector.1(c.a, pos, args$window.size.voting.X / 2)\n",
    "        c.b <- correct.vector.1(c.b, pos, args$window.size.voting.X / 2)\n",
    "      }\n",
    "      parsed.flow[[chr]][,i] <- paste0(c.a, '|', c.b)\n",
    "      is.corrected[[chr]][,(i*2)-1] <- a != c.a\n",
    "      is.corrected[[chr]][,i*2] <- b != c.b\n",
    "    }\n",
    "  }\n",
    "  return(list(parsed.flow = parsed.flow, is.corrected = is.corrected))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b94fa428-487b-45e6-9d43-d921ec17925c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# functions for making plot\n",
    "\n",
    "get.haplo.profiles <- function(){\n",
    "  get.breaks <- function(f, pos){\n",
    "    changes <- cumsum(rle(f)$lengths)\n",
    "    colors <- letter.colors[rle(f)$values]\n",
    "    breakpoints <- sapply(changes, function(x) pos[x] + (pos[x+1] - pos[x]) / 2)\n",
    "    breakpoints <- breakpoints[-length(breakpoints)]\n",
    "    return(list(breakpoints = breakpoints, colors = colors))\n",
    "  }\n",
    "  \n",
    "  filter.region <- function(frame, chr, whole.chromosome = F){\n",
    "    if (length(args$regions)){\n",
    "      for (region in args$regions){\n",
    "        r.split <- strsplit(region, ':')[[1]]\n",
    "        if (chr == r.split[1]){\n",
    "          if (whole.chromosome) return(frame)\n",
    "          s <- as.numeric(strsplit(r.split[2],'-')[[1]][1])\n",
    "          s <- s - args$regions.flanking.size\n",
    "          e <- as.numeric(strsplit(r.split[2],'-')[[1]][2])\n",
    "          e <- e + args$regions.flanking.size\n",
    "          return(frame[frame$x >= s & frame$x <= e,])\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    return(frame[-(1:nrow(frame)),])\n",
    "  }\n",
    "  \n",
    "  add.marker.and.trace <- function(fig, y, f = 'f1', mar = 'y-down-open', symbol.offset = .3){\n",
    "    for (i in 1:(length(breaks[[s]][[f]])-1)){\n",
    "      if (i != 1){\n",
    "        fig <- add_markers(fig, x = breaks[[s]][[f]][i], y = rep(y + symbol.offset, 2),\n",
    "                           name = NULL, hoverinfo = \"none\",\n",
    "                           marker = list(symbol = mar, color = colors[length(letters) + 1],\n",
    "                                         size = args$dot.factor * 8),\n",
    "                           type='scatter', mode='marker', inherit = F)\n",
    "      }\n",
    "      fig <- add_trace(fig, x = c(breaks[[s]][[f]][i], breaks[[s]][[f]][i+1]), y = rep(y, 2),\n",
    "                       name = NULL, hoverinfo = \"none\",\n",
    "                       line = list(color = breaks[[s]][[paste0(f, 'cols')]][i], width = args$dot.factor * 2),\n",
    "                       type='scatter', mode='lines', inherit = F)\n",
    "    }\n",
    "    return(fig)\n",
    "  }\n",
    "  \n",
    "  chr.lengths <- sapply(chrs, function(x) max(vcfs.filtered2[[1]]$POS[vcfs.filtered2[[1]]$CHROM == x]))\n",
    "  \n",
    "  haplo.profiles <- list()\n",
    "  for (c in chrs){\n",
    "    haplo.frame <- matrix(nrow = 0, ncol = 9)\n",
    "    annot.list <- list()\n",
    "    breaks <- list()\n",
    "    for (s in args$samples.no.u){\n",
    "      x <- map.list[[c]]$pos\n",
    "      f1 <- sapply(strsplit(parsed.flow[[c]][,which(args$samples.no.u == s)], '|', fixed = T), function(x) x[1])\n",
    "      f2 <- sapply(strsplit(parsed.flow[[c]][,which(args$samples.no.u == s)], '|', fixed = T), function(x) x[2])\n",
    "      g1 <- sapply(strsplit(parsed.geno[[c]][,which(args$samples.no.u == s)], '|', fixed = T), function(x) x[1])\n",
    "      g2 <- sapply(strsplit(parsed.geno[[c]][,which(args$samples.no.u == s)], '|', fixed = T), function(x) x[2])\n",
    "      c1 <- is.corrected[[c]][,which(args$samples.no.u == s) * 2 - 1]\n",
    "      c2 <- is.corrected[[c]][,which(args$samples.no.u == s) * 2]\n",
    "      breaks[[s]]$f1 <- c(x[1], get.breaks(f1, x)$breakpoints, x[length(x)])\n",
    "      breaks[[s]]$f1cols <- get.breaks(f1, x)$colors\n",
    "      breaks[[s]]$f2 <- c(x[1], get.breaks(f2, x)$breakpoints, x[length(x)])\n",
    "      breaks[[s]]$f2cols <- get.breaks(f2, x)$colors\n",
    "      symbol <- rep(1, length(c1) * 2)\n",
    "      symbol[c(c1,c2)] <- 0\n",
    "      symbol[c(g1,g2) == 'NA'] <- 101 ## will show nothing\n",
    "      id1 <- paste0(c, ':', map.list[[c]]$pos.out, ' (', g1, ')')\n",
    "      id2 <- paste0(c, ':', map.list[[c]]$pos.out, ' (', g2, ')')\n",
    "      y1 <- length(args$samples.no.u) * 3 - which(args$samples.no.u == s) * 3 + 1\n",
    "      y2 <- length(args$samples.no.u) * 3 - which(args$samples.no.u == s) * 3\n",
    "      haplo.frame.sub <- data.frame(c(x, x), c(rep(y1, length(x)), rep(y2, length(x))),\n",
    "                                    c(id1, id2), c(letter.colors[f1],  letter.colors[f2]), symbol,\n",
    "                                    stringsAsFactors = F, c(rep(which(args$samples.no.u == s) * 2 - 1, length(x)),\n",
    "                                                            rep(which(args$samples.no.u == s) * 2, length(x))))\n",
    "      annot.list[[which(args$samples.no.u == s)]] <- list(x = 1, y = y1 + 1,\n",
    "                                                          text = args$samples.out[args$sample.ids == s], showarrow = F)\n",
    "      if (all(c(g1, g2) == 'NA')) next\n",
    "      haplo.frame <- rbind(haplo.frame, haplo.frame.sub)\n",
    "    }\n",
    "    \n",
    "    ## raw data points\n",
    "    \n",
    "    colnames(haplo.frame) <- c('x', 'y', 'id', 'col', 'symbol', 'name')\n",
    "    haplo.frame.for.plotly <- haplo.frame\n",
    "    if (args$keep.chromosomes.only) haplo.frame.for.plotly <- filter.region(haplo.frame, c, whole.chromosome = T)\n",
    "    if (args$keep.regions.only) haplo.frame.for.plotly <- filter.region(haplo.frame, c)\n",
    "    p <- plot_ly(haplo.frame.for.plotly, x =~x, y =~y, text =~id, name=~name, marker = list(color=~col,\n",
    "                                                                                            symbol=~symbol,\n",
    "                                                                                            size = args$dot.factor * 4,\n",
    "                                                                                            line=list(color=~col)),\n",
    "                 type = 'scattergl', mode = 'markers', hoverinfo = 'text', height = 900 * length(args$samples.no.u),\n",
    "                 hoverlabel=list(bgcolor=~col))\n",
    "    \n",
    "    ## layout\n",
    "    \n",
    "    p <- p %>% layout(xaxis = list(title = list(text=c, standoff=10), showticklabels = F,\n",
    "                                   zeroline = F, showgrid = F), showlegend = F,\n",
    "                      yaxis = list(title = '',showticklabels = F, zeroline = F, showgrid = F,\n",
    "                                   fixedrange = T, range = c(-1, length(args$samples.no.u) * 3 + 1)),\n",
    "                      annotations = annot.list)\n",
    "    \n",
    "    ## add traces and recombination points\n",
    "    \n",
    "    for (s in args$samples.no.u){\n",
    "      y2 <- length(args$samples.no.u) * 3 - which(args$samples.no.u == s) * 3\n",
    "      p <- add.marker.and.trace(p, length(args$samples.no.u) * 3 - which(args$samples.no.u == s) * 3 + 1)\n",
    "      p <- add.marker.and.trace(p, length(args$samples.no.u) * 3 - which(args$samples.no.u == s) * 3, 'f2', 'y-up-open', symbol.offset = -.3)\n",
    "    }\n",
    "    \n",
    "    ## add regions\n",
    "    \n",
    "    if (length(args$regions)){\n",
    "      for (region in args$regions){\n",
    "        if (c == strsplit(region, ':')[[1]][1]){\n",
    "          tmp <- 1 ; names(tmp) <- c\n",
    "          p <- mark.region(p, tmp, c(-.5, length(args$samples.no.u) * 3 + 1), region, chr.lengths)\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    haplo.profiles[[c]] <- p\n",
    "    break\n",
    "  }\n",
    "  return (haplo.profiles)\n",
    "}\n",
    "\n",
    "get.breaks <- function(f, pos){\n",
    "    changes <- cumsum(rle(f)$lengths)\n",
    "    colors <- letter.colors[rle(f)$values]\n",
    "    breakpoints <- sapply(changes, function(x) pos[x] + (pos[x+1] - pos[x]) / 2)\n",
    "    breakpoints <- breakpoints[-length(breakpoints)]\n",
    "    return(list(breakpoints = breakpoints, colors = colors))\n",
    "}\n",
    "\n",
    "add.marker.and.trace <- function(fig, y, f = 'f1', mar = 'y-down-open', symbol.offset = .3){\n",
    "for (i in 1:(length(breaks[[s]][[f]])-1)){\n",
    "  if (i != 1){\n",
    "    fig <- add_markers(fig, x = breaks[[s]][[f]][i], y = rep(y + symbol.offset, 2),\n",
    "                       name = NULL, hoverinfo = \"none\",\n",
    "                       marker = list(symbol = mar, color = colors[length(letters) + 1],\n",
    "                                     size = args$dot.factor * 8),\n",
    "                       type='scatter', mode='marker', inherit = F)\n",
    "  }\n",
    "  fig <- add_trace(fig, x = c(breaks[[s]][[f]][i], breaks[[s]][[f]][i+1]), y = rep(y, 2),\n",
    "                   name = NULL, hoverinfo = \"none\",\n",
    "                   line = list(color = breaks[[s]][[paste0(f, 'cols')]][i], width = args$dot.factor * 2),\n",
    "                   type='scatter', mode='lines', inherit = F)\n",
    "}\n",
    "return(fig)\n",
    "}\n",
    "\n",
    "mark.region <- function(fig, chr.cs, ylim, region, chr.lengths, plot.flanks = T){\n",
    "  c <- strsplit(region, ':')[[1]][1]\n",
    "  s <- as.numeric(strsplit(strsplit(region, ':')[[1]][2], '-')[[1]][1]) ; s <- max(s, 1)\n",
    "  flank.s <- s - args$regions.flanking.size ; flank.s <- max(flank.s, 1)\n",
    "  e <- as.numeric(strsplit(strsplit(region, ':')[[1]][2], '-')[[1]][2]) ; e <- min(e, chr.lengths[c])\n",
    "  flank.e <- e + args$regions.flanking.size ; flank.e <- min(flank.e, chr.lengths[c])\n",
    "  \n",
    "  region.start <- paste0(c, ':', scales::comma(s, accuracy = 1))\n",
    "  region.end <- paste0(c, ':', scales::comma(e, accuracy = 1))\n",
    "  flank.start <- paste0(c, ':', scales::comma(flank.s, accuracy = 1))\n",
    "  flank.end <- paste0(c, ':', scales::comma(flank.e, accuracy = 1))\n",
    "  \n",
    "  add.trace <- function(fig, xs, text, size = 1/2){\n",
    "    fig <- fig %>% add_trace(x = xs, y = ylim, name = 'region',\n",
    "                             hoverinfo = \"text\", line = list(color = colors[length(letters) + 1], width = args$dot.factor * size),\n",
    "                             text = text, type='scatter', marker = list(color = colors[length(letters) + 1], size = .1),\n",
    "                             mode='lines+markers', hoverlabel=list(bgcolor=colors[length(letters) + 1]), inherit = F)\n",
    "    return(fig)\n",
    "  }\n",
    "  \n",
    "  fig <- add.trace(fig, c(chr.cs[c] + s, chr.cs[c] + s), paste0(region.start, ' (region start)'))\n",
    "  fig <- add.trace(fig, c(chr.cs[c] + e, chr.cs[c] + e), paste0(region.end, ' (region end)'))\n",
    "  \n",
    "  if (plot.flanks){\n",
    "    fig <- add.trace(fig, rep(chr.cs[c] + flank.s, 2), paste0(flank.start, ' (flank start)'), 1/4)\n",
    "    fig <- add.trace(fig, rep(chr.cs[c] + flank.e, 2), paste0(flank.end, ' (flank end)'), 1/4)\n",
    "  }\n",
    "  return(fig)\n",
    "}\n",
    "\n",
    "add.cytoband <- function(fig, chr, y, line.width = 4){\n",
    "  s <- c()\n",
    "  names <- c()\n",
    "  cols <- c()\n",
    "  for (i in 1:length(cytobands[[chr]])){\n",
    "    xs <- c(cytobands[[chr]][[i]]$start, cytobands[[chr]][[i]]$end)\n",
    "    col <- c('lightgrey', 'darkgrey')[[i %% 2 + 1]]\n",
    "    if (any(sapply(c('acen', 'gvar', 'stalk'), function(x) grepl(x, cytobands[[chr]][[i]]$name)))) col <- 'black'\n",
    "    fig <- add_trace(fig, x = xs, y = rep(y, 2),\n",
    "                     name = NULL, hoverinfo = \"none\",\n",
    "                     line = list(color = col, width = args$dot.factor * line.width),\n",
    "                     type='scatter', mode='lines', inherit = F)\n",
    "    seq <- seq(xs[1], xs[2], 10000000)\n",
    "    s <- c(s, seq)\n",
    "    names <- c(names, paste0(cytobands[[chr]][[i]]$name, ' (',\n",
    "                             scales::comma(seq, accuracy = 1), ')'))\n",
    "    cols <- c(cols, rep(col, length(seq)))\n",
    "  }\n",
    "  \n",
    "  fig <- add_markers(fig, x = s, y = rep(y, length(s)), text=names, name = NULL,\n",
    "                     hoverinfo='text', marker = list(color = cols, size = args$dot.factor * .1),\n",
    "                     type='scatter', mode='marker', inherit = F)\n",
    "  return(fig)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "128bb38d-76a5-41e4-bb9d-e6908ddbb0bc",
   "metadata": {
    "jupyter": {
     "source_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "get.empty <- function(){\n",
    "  p <- plot_ly(x = 0, y = 0, name = NULL, type = 'scatter', mode = 'markers',\n",
    "                marker = list(color = 'white', size = .01), hoverinfo='skip')\n",
    "  p <- p %>% layout(xaxis = list(title = '', showticklabels = F, zeroline = F, showgrid = F, fixedrange = T),\n",
    "                    yaxis = list(title = '', showticklabels = F, zeroline = F, showgrid = F, fixedrange = T),\n",
    "                    showlegend = F)\n",
    "  return(p)\n",
    "}\n",
    "\n",
    "\n",
    "do.subplot <- function(plot.list, ncol = 1, panning = .03, margin = .02, overridehovermode = NULL){\n",
    "  nrow <- ceiling(length(plot.list) / ncol)\n",
    "  \n",
    "  plot.list.ap <- lapply(1:(ncol + 2), function(x) get.empty())\n",
    "  for (i in 1:nrow){\n",
    "    plot.list.ap <- c(plot.list.ap, lapply(1, function(x) get.empty()))\n",
    "    s <- (1 + (i - 1) * ncol)\n",
    "    e <- (ncol + (i - 1) * ncol)\n",
    "    real.e <- length(plot.list)\n",
    "    plot.list.ap <- c(plot.list.ap, plot.list[s:min(real.e, e)])\n",
    "    plot.list.ap <- c(plot.list.ap, lapply(1, function(x) get.empty()))\n",
    "  }\n",
    "  plot.list.ap <- c(plot.list.ap, lapply((length(plot.list.ap)+1):((ncol+2)*(nrow+2)), function(x) get.empty()))\n",
    "  \n",
    "  heights <- c(panning, rep((1 - 2*panning) / nrow, nrow), panning)\n",
    "  widths <- c(panning, rep((1 - 2*panning) / ncol, ncol), panning)\n",
    "  this.subplot <- subplot(plot.list.ap, nrows=nrow+2, margin=margin, titleY=T, titleX=T,\n",
    "                          heights=heights, widths = widths)\n",
    "  if (!is.null(overridehovermode)) this.subplot$x$layout$hovermode <- overridehovermode\n",
    "  return(this.subplot)\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce29c312-4e4e-4146-8ce8-5e26666bf7fb",
   "metadata": {},
   "source": [
    "# Read the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c38aee6c-ddad-488f-be1c-ccb273c74f13",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: Argument --visualization does not exist.\n",
      "Selected parameters ...\n",
      "  ... vcf.file: /home/jovyan/work/hopla/inkyun/data/Proband_16_09932-gatk-haplotype-joint-annotated.vcf.gz\n",
      "  ... sample.ids: U8,D2301776,D2303583,K2300654,K2300655,K2300656,K2300657\n",
      "  ... father.ids: NA,NA,U8,U8,U8,U8,U8\n",
      "  ... mother.ids: NA,NA,D2301776,D2301776,D2301776,D2301776,D2301776\n",
      "  ... genders: M,F,M,NA,NA,NA,NA\n",
      "  ... run.merlin: TRUE\n",
      "  ... cytoband.file: /home/jovyan/work/hopla/data/cytoband.hg38.txt\n",
      "  ... dp.hard.limit.ids: D2301776,D2303583\n",
      "  ... dp.hard.limit: 10\n",
      "  ... af.hard.limit.ids: D2301776,D2303583\n",
      "  ... af.hard.limit: 0.25\n",
      "  ... dp.soft.limit.ids: K2300654,K2300655,K2300656,K2300657\n",
      "  ... dp.soft.limit: 10\n",
      "  ... regions: chrX:53176283-53225422\n",
      "  ... carrier.ids: D2301776\n",
      "  ... affected.ids: D2303583\n",
      "  ... info: Disease:Intellectual developmental disorder KDM5C,Inheritance:XLR,Sequencing note:\n",
      "  ... baf.ids: K2300654,K2300655,K2300656,K2300657\n",
      "  ... merlin.model: best\n",
      "  ... min.seg.var: 5\n",
      "  ... min.seg.var.X: 15\n",
      "  ... window.size.voting: 1e+07\n",
      "  ... window.size.voting.X: 1e+07\n",
      "  ... keep.chromosomes.only: TRUE\n",
      "  ... keep.regions.only: FALSE\n",
      "  ... concordance.table: TRUE\n",
      "  ... out.dir: /home/jovyan/work/hopla/inkyun/haplotyping_demo/\n",
      "  ... fam.id: Proband_16_09932_A\n",
      "  ... X.cutoff: 1.5\n",
      "  ... Y.cutoff: 0.6\n",
      "  ... window.size: 1e+06\n",
      "  ... regions.flanking.size: 1e+06\n",
      "  ... limit.baf.to.P: FALSE\n",
      "  ... limit.pm.to.P: TRUE\n",
      "  ... value.of.P: 0.15\n",
      "  ... color.palette: Paired\n",
      "  ... dot.factor: 2\n",
      "  ... self.contained: TRUE\n",
      "  ... cairo: FALSE\n",
      "  ... visualization: F\n"
     ]
    }
   ],
   "source": [
    "setting <- '/home/jovyan/work/hopla/data/setting_haplotyping_demo.txt'\n",
    "\n",
    "args <- get.file.args(setting)\n",
    "args <- post.process.args(args)\n",
    "\n",
    "dir.create(args$out.dir, showWarnings = F, recursive = T)\n",
    "args$data.dir <- dirname(args$vcf.file)\n",
    "\n",
    "chrs <- paste0('chr', c(1:22, 'X'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1c7e0f33-635f-4adf-8f3c-51e002b16f83",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load vcfs\n",
    "\n",
    "vcfs_merged <- read.csv(file.path(args$data.dir, 'vcfs.csv'))\n",
    "vcfs_merged <- vcfs_merged[,-1]\n",
    "vcfs <- list()\n",
    "\n",
    "for (i in 1:length(vcfs_merged)) {\n",
    "    column_name = colnames(vcfs_merged)[i]\n",
    "    sample_name <- str_split(column_name, '\\\\.', n=2)[[1]]\n",
    "    sampleID <- sample_name[1]\n",
    "    col <- sample_name[2]\n",
    "  \n",
    "    if (is.null(vcfs[[sampleID]])) {\n",
    "        vcfs[[sampleID]] <- data.frame(vcfs_merged[[column_name]])\n",
    "    } else {\n",
    "        vcfs[[sampleID]] <- cbind(vcfs[[sampleID]], vcfs_merged[[column_name]])\n",
    "    }\n",
    "\n",
    "    if (i%%12 == 0) {\n",
    "        colnames(vcfs[[sampleID]])[12] <- sample_name[2]\n",
    "    } else {\n",
    "        colnames(vcfs[[sampleID]])[i%%12] <- sample_name[2]\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c7f8c7f3-afa1-4c77-baa3-450300c7a901",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract vcfs.filtered and vcfs.filtered2 from vcfs\n",
    "\n",
    "vcfs.filtered <- list()\n",
    "vcfs.filtered2 <- list()\n",
    "\n",
    "for (i in 1:length(vcfs)) {\n",
    "    last.col.num <- length(vcfs[[1]])-2\n",
    "    vcfs.filtered[[i]] <- vcfs[[i]][vcfs[[i]]$On_filter1 == 1, ][1:last.col.num]\n",
    "    vcfs.filtered2[[i]] <- vcfs[[i]][vcfs[[i]]$On_filter2 == 1, ][1:last.col.num]\n",
    "    vcfs[[i]] <- vcfs[[i]][1:last.col.num]\n",
    "    colnames(vcfs[[i]])[ncol(vcfs[[i]])] <- \"POS.OUT\"\n",
    "\n",
    "    hom.ref <- which(vcfs.filtered[[i]]$GT == '0/0')\n",
    "    het.alt <- which(vcfs.filtered[[i]]$GT == '0/1')\n",
    "    hom.alt <- which(vcfs.filtered[[i]]$GT == '1/1')\n",
    "    \n",
    "    vcfs.filtered[[i]]$GENO <- 'N/N'\n",
    "    vcfs.filtered[[i]]$GENO[hom.ref] <- paste0(vcfs.filtered[[i]]$REF, '/', vcfs.filtered[[i]]$REF)[hom.ref]\n",
    "    vcfs.filtered[[i]]$GENO[het.alt] <- paste0(vcfs.filtered[[i]]$REF, '/', vcfs.filtered[[i]]$ALT)[het.alt]\n",
    "    vcfs.filtered[[i]]$GENO[hom.alt] <- paste0(vcfs.filtered[[i]]$ALT, '/', vcfs.filtered[[i]]$ALT)[hom.alt]\n",
    "\n",
    "    hom.ref <- which(vcfs.filtered2[[i]]$GT == '0/0')\n",
    "    het.alt <- which(vcfs.filtered2[[i]]$GT == '0/1')\n",
    "    hom.alt <- which(vcfs.filtered2[[i]]$GT == '1/1')\n",
    "    \n",
    "    vcfs.filtered2[[i]]$GENO <- 'N/N'\n",
    "    vcfs.filtered2[[i]]$GENO[hom.ref] <- paste0(vcfs.filtered2[[i]]$REF, '/', vcfs.filtered2[[i]]$REF)[hom.ref]\n",
    "    vcfs.filtered2[[i]]$GENO[het.alt] <- paste0(vcfs.filtered2[[i]]$REF, '/', vcfs.filtered2[[i]]$ALT)[het.alt]\n",
    "    vcfs.filtered2[[i]]$GENO[hom.alt] <- paste0(vcfs.filtered2[[i]]$ALT, '/', vcfs.filtered2[[i]]$ALT)[hom.alt]\n",
    "\n",
    "\n",
    "    \n",
    "}\n",
    "\n",
    "names(vcfs.filtered) <- names(vcfs)\n",
    "names(vcfs.filtered2) <- names(vcfs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bd5be5bd-f99e-459a-a9b1-c9579f54aa66",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'D2301776'</li><li>'D2303583'</li><li>'K2300654'</li><li>'K2300655'</li><li>'K2300656'</li><li>'K2300657'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'D2301776'\n",
       "\\item 'D2303583'\n",
       "\\item 'K2300654'\n",
       "\\item 'K2300655'\n",
       "\\item 'K2300656'\n",
       "\\item 'K2300657'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'D2301776'\n",
       "2. 'D2303583'\n",
       "3. 'K2300654'\n",
       "4. 'K2300655'\n",
       "5. 'K2300656'\n",
       "6. 'K2300657'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"D2301776\" \"D2303583\" \"K2300654\" \"K2300655\" \"K2300656\" \"K2300657\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>CHROM</th><th scope=col>POS</th><th scope=col>ID</th><th scope=col>REF</th><th scope=col>ALT</th><th scope=col>GT</th><th scope=col>AD</th><th scope=col>DP</th><th scope=col>AF</th><th scope=col>POS.OUT</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>chr1</td><td>48186</td><td>id1</td><td>T</td><td>G</td><td>0/1</td><td> 4</td><td> 4</td><td>0.500</td><td>48,186</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>chr1</td><td>51431</td><td>id2</td><td>C</td><td>T</td><td>0/0</td><td> 0</td><td> 0</td><td>   NA</td><td>51,431</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>chr1</td><td>51476</td><td>id3</td><td>T</td><td>C</td><td>0/0</td><td> 0</td><td> 0</td><td>   NA</td><td>51,476</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>chr1</td><td>60796</td><td>id6</td><td>C</td><td>G</td><td>0/0</td><td> 0</td><td> 0</td><td>   NA</td><td>60,796</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>chr1</td><td>63671</td><td>id7</td><td>G</td><td>A</td><td>0/1</td><td>19</td><td>19</td><td>0.421</td><td>63,671</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>chr1</td><td>63697</td><td>id8</td><td>T</td><td>C</td><td>0/1</td><td>19</td><td>19</td><td>0.421</td><td>63,697</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 10\n",
       "\\begin{tabular}{r|llllllllll}\n",
       "  & CHROM & POS & ID & REF & ALT & GT & AD & DP & AF & POS.OUT\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & chr1 & 48186 & id1 & T & G & 0/1 &  4 &  4 & 0.500 & 48,186\\\\\n",
       "\t2 & chr1 & 51431 & id2 & C & T & 0/0 &  0 &  0 &    NA & 51,431\\\\\n",
       "\t3 & chr1 & 51476 & id3 & T & C & 0/0 &  0 &  0 &    NA & 51,476\\\\\n",
       "\t4 & chr1 & 60796 & id6 & C & G & 0/0 &  0 &  0 &    NA & 60,796\\\\\n",
       "\t5 & chr1 & 63671 & id7 & G & A & 0/1 & 19 & 19 & 0.421 & 63,671\\\\\n",
       "\t6 & chr1 & 63697 & id8 & T & C & 0/1 & 19 & 19 & 0.421 & 63,697\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 10\n",
       "\n",
       "| <!--/--> | CHROM &lt;chr&gt; | POS &lt;int&gt; | ID &lt;chr&gt; | REF &lt;chr&gt; | ALT &lt;chr&gt; | GT &lt;chr&gt; | AD &lt;int&gt; | DP &lt;int&gt; | AF &lt;dbl&gt; | POS.OUT &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | chr1 | 48186 | id1 | T | G | 0/1 |  4 |  4 | 0.500 | 48,186 |\n",
       "| 2 | chr1 | 51431 | id2 | C | T | 0/0 |  0 |  0 |    NA | 51,431 |\n",
       "| 3 | chr1 | 51476 | id3 | T | C | 0/0 |  0 |  0 |    NA | 51,476 |\n",
       "| 4 | chr1 | 60796 | id6 | C | G | 0/0 |  0 |  0 |    NA | 60,796 |\n",
       "| 5 | chr1 | 63671 | id7 | G | A | 0/1 | 19 | 19 | 0.421 | 63,671 |\n",
       "| 6 | chr1 | 63697 | id8 | T | C | 0/1 | 19 | 19 | 0.421 | 63,697 |\n",
       "\n"
      ],
      "text/plain": [
       "  CHROM POS   ID  REF ALT GT  AD DP AF    POS.OUT\n",
       "1 chr1  48186 id1 T   G   0/1  4  4 0.500 48,186 \n",
       "2 chr1  51431 id2 C   T   0/0  0  0    NA 51,431 \n",
       "3 chr1  51476 id3 T   C   0/0  0  0    NA 51,476 \n",
       "4 chr1  60796 id6 C   G   0/0  0  0    NA 60,796 \n",
       "5 chr1  63671 id7 G   A   0/1 19 19 0.421 63,671 \n",
       "6 chr1  63697 id8 T   C   0/1 19 19 0.421 63,697 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>1624518</li><li>10</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 1624518\n",
       "\\item 10\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 1624518\n",
       "2. 10\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 1624518      10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vcfs %>% names\n",
    "vcfs[[1]] %>% head\n",
    "vcfs[[1]] %>% dim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "73385ebb-13fd-4651-a785-62db403fc405",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'D2301776'</li><li>'D2303583'</li><li>'K2300654'</li><li>'K2300655'</li><li>'K2300656'</li><li>'K2300657'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'D2301776'\n",
       "\\item 'D2303583'\n",
       "\\item 'K2300654'\n",
       "\\item 'K2300655'\n",
       "\\item 'K2300656'\n",
       "\\item 'K2300657'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'D2301776'\n",
       "2. 'D2303583'\n",
       "3. 'K2300654'\n",
       "4. 'K2300655'\n",
       "5. 'K2300656'\n",
       "6. 'K2300657'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"D2301776\" \"D2303583\" \"K2300654\" \"K2300655\" \"K2300656\" \"K2300657\""
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 11</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>CHROM</th><th scope=col>POS</th><th scope=col>ID</th><th scope=col>REF</th><th scope=col>ALT</th><th scope=col>GT</th><th scope=col>AD</th><th scope=col>DP</th><th scope=col>AF</th><th scope=col>POS.out</th><th scope=col>GENO</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>13</th><td>chr1</td><td> 84244</td><td>id16</td><td>A</td><td>C</td><td>0/0</td><td>14</td><td>14</td><td>0.000</td><td>84,244 </td><td>A/A</td></tr>\n",
       "\t<tr><th scope=row>20</th><td>chr1</td><td>261753</td><td>id26</td><td>G</td><td>A</td><td>0/1</td><td>46</td><td>46</td><td>0.348</td><td>261,753</td><td>G/A</td></tr>\n",
       "\t<tr><th scope=row>23</th><td>chr1</td><td>266225</td><td>id30</td><td>C</td><td>A</td><td>0/1</td><td>28</td><td>28</td><td>0.714</td><td>266,225</td><td>C/A</td></tr>\n",
       "\t<tr><th scope=row>37</th><td>chr1</td><td>665098</td><td>id50</td><td>G</td><td>A</td><td>0/0</td><td>10</td><td>10</td><td>0.000</td><td>665,098</td><td>G/G</td></tr>\n",
       "\t<tr><th scope=row>58</th><td>chr1</td><td>852019</td><td>id78</td><td>G</td><td>T</td><td>1/1</td><td>31</td><td>31</td><td>1.000</td><td>852,019</td><td>T/T</td></tr>\n",
       "\t<tr><th scope=row>59</th><td>chr1</td><td>852464</td><td>id79</td><td>C</td><td>T</td><td>1/1</td><td>18</td><td>18</td><td>1.000</td><td>852,464</td><td>T/T</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 11\n",
       "\\begin{tabular}{r|lllllllllll}\n",
       "  & CHROM & POS & ID & REF & ALT & GT & AD & DP & AF & POS.out & GENO\\\\\n",
       "  & <chr> & <int> & <chr> & <chr> & <chr> & <chr> & <int> & <int> & <dbl> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t13 & chr1 &  84244 & id16 & A & C & 0/0 & 14 & 14 & 0.000 & 84,244  & A/A\\\\\n",
       "\t20 & chr1 & 261753 & id26 & G & A & 0/1 & 46 & 46 & 0.348 & 261,753 & G/A\\\\\n",
       "\t23 & chr1 & 266225 & id30 & C & A & 0/1 & 28 & 28 & 0.714 & 266,225 & C/A\\\\\n",
       "\t37 & chr1 & 665098 & id50 & G & A & 0/0 & 10 & 10 & 0.000 & 665,098 & G/G\\\\\n",
       "\t58 & chr1 & 852019 & id78 & G & T & 1/1 & 31 & 31 & 1.000 & 852,019 & T/T\\\\\n",
       "\t59 & chr1 & 852464 & id79 & C & T & 1/1 & 18 & 18 & 1.000 & 852,464 & T/T\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 11\n",
       "\n",
       "| <!--/--> | CHROM &lt;chr&gt; | POS &lt;int&gt; | ID &lt;chr&gt; | REF &lt;chr&gt; | ALT &lt;chr&gt; | GT &lt;chr&gt; | AD &lt;int&gt; | DP &lt;int&gt; | AF &lt;dbl&gt; | POS.out &lt;chr&gt; | GENO &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 13 | chr1 |  84244 | id16 | A | C | 0/0 | 14 | 14 | 0.000 | 84,244  | A/A |\n",
       "| 20 | chr1 | 261753 | id26 | G | A | 0/1 | 46 | 46 | 0.348 | 261,753 | G/A |\n",
       "| 23 | chr1 | 266225 | id30 | C | A | 0/1 | 28 | 28 | 0.714 | 266,225 | C/A |\n",
       "| 37 | chr1 | 665098 | id50 | G | A | 0/0 | 10 | 10 | 0.000 | 665,098 | G/G |\n",
       "| 58 | chr1 | 852019 | id78 | G | T | 1/1 | 31 | 31 | 1.000 | 852,019 | T/T |\n",
       "| 59 | chr1 | 852464 | id79 | C | T | 1/1 | 18 | 18 | 1.000 | 852,464 | T/T |\n",
       "\n"
      ],
      "text/plain": [
       "   CHROM POS    ID   REF ALT GT  AD DP AF    POS.out GENO\n",
       "13 chr1   84244 id16 A   C   0/0 14 14 0.000 84,244  A/A \n",
       "20 chr1  261753 id26 G   A   0/1 46 46 0.348 261,753 G/A \n",
       "23 chr1  266225 id30 C   A   0/1 28 28 0.714 266,225 C/A \n",
       "37 chr1  665098 id50 G   A   0/0 10 10 0.000 665,098 G/G \n",
       "58 chr1  852019 id78 G   T   1/1 31 31 1.000 852,019 T/T \n",
       "59 chr1  852464 id79 C   T   1/1 18 18 1.000 852,464 T/T "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>131955</li><li>11</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 131955\n",
       "\\item 11\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 131955\n",
       "2. 11\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 131955     11"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "vcfs.filtered %>% names\n",
    "vcfs.filtered[[1]] %>% head\n",
    "vcfs.filtered[[1]] %>% dim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e49ff706-f267-448b-9f6f-d78839ae4c77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load parsed.flow\n",
    "\n",
    "parsed.flow.raw <- read.csv(file.path(args$data.dir, 'parsed_flow.csv'))\n",
    "parsed.flow <- list()\n",
    "\n",
    "for (i in 1:23) {\n",
    "    chr = paste0('chr', i)\n",
    "    parsed.flow[[i]] <- parsed.flow.raw[parsed.flow.raw$V7 == chr, ][,-7]\n",
    "}\n",
    "\n",
    "names(parsed.flow) <- chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "65a7f9b0-eb3b-45d4-8813-ad9ffb60f06f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'chr1'</li><li>'chr2'</li><li>'chr3'</li><li>'chr4'</li><li>'chr5'</li><li>'chr6'</li><li>'chr7'</li><li>'chr8'</li><li>'chr9'</li><li>'chr10'</li><li>'chr11'</li><li>'chr12'</li><li>'chr13'</li><li>'chr14'</li><li>'chr15'</li><li>'chr16'</li><li>'chr17'</li><li>'chr18'</li><li>'chr19'</li><li>'chr20'</li><li>'chr21'</li><li>'chr22'</li><li>'chrX'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'chr1'\n",
       "\\item 'chr2'\n",
       "\\item 'chr3'\n",
       "\\item 'chr4'\n",
       "\\item 'chr5'\n",
       "\\item 'chr6'\n",
       "\\item 'chr7'\n",
       "\\item 'chr8'\n",
       "\\item 'chr9'\n",
       "\\item 'chr10'\n",
       "\\item 'chr11'\n",
       "\\item 'chr12'\n",
       "\\item 'chr13'\n",
       "\\item 'chr14'\n",
       "\\item 'chr15'\n",
       "\\item 'chr16'\n",
       "\\item 'chr17'\n",
       "\\item 'chr18'\n",
       "\\item 'chr19'\n",
       "\\item 'chr20'\n",
       "\\item 'chr21'\n",
       "\\item 'chr22'\n",
       "\\item 'chrX'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'chr1'\n",
       "2. 'chr2'\n",
       "3. 'chr3'\n",
       "4. 'chr4'\n",
       "5. 'chr5'\n",
       "6. 'chr6'\n",
       "7. 'chr7'\n",
       "8. 'chr8'\n",
       "9. 'chr9'\n",
       "10. 'chr10'\n",
       "11. 'chr11'\n",
       "12. 'chr12'\n",
       "13. 'chr13'\n",
       "14. 'chr14'\n",
       "15. 'chr15'\n",
       "16. 'chr16'\n",
       "17. 'chr17'\n",
       "18. 'chr18'\n",
       "19. 'chr19'\n",
       "20. 'chr20'\n",
       "21. 'chr21'\n",
       "22. 'chr22'\n",
       "23. 'chrX'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"chr1\"  \"chr2\"  \"chr3\"  \"chr4\"  \"chr5\"  \"chr6\"  \"chr7\"  \"chr8\"  \"chr9\" \n",
       "[10] \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\"\n",
       "[19] \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>V6</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A|B</td><td>A|C</td><td>B|D</td><td>A|D</td><td>B|D</td><td>A|C</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A|B</td><td>A|C</td><td>B|D</td><td>A|D</td><td>B|D</td><td>A|C</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>A|B</td><td>A|C</td><td>B|D</td><td>A|D</td><td>B|D</td><td>A|C</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>A|B</td><td>A|C</td><td>B|D</td><td>A|D</td><td>B|D</td><td>A|C</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>A|B</td><td>A|C</td><td>B|D</td><td>A|D</td><td>B|D</td><td>A|C</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>A|B</td><td>A|C</td><td>B|D</td><td>A|D</td><td>B|D</td><td>A|C</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & V1 & V2 & V3 & V4 & V5 & V6\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & A\\textbar{}B & A\\textbar{}C & B\\textbar{}D & A\\textbar{}D & B\\textbar{}D & A\\textbar{}C\\\\\n",
       "\t2 & A\\textbar{}B & A\\textbar{}C & B\\textbar{}D & A\\textbar{}D & B\\textbar{}D & A\\textbar{}C\\\\\n",
       "\t3 & A\\textbar{}B & A\\textbar{}C & B\\textbar{}D & A\\textbar{}D & B\\textbar{}D & A\\textbar{}C\\\\\n",
       "\t4 & A\\textbar{}B & A\\textbar{}C & B\\textbar{}D & A\\textbar{}D & B\\textbar{}D & A\\textbar{}C\\\\\n",
       "\t5 & A\\textbar{}B & A\\textbar{}C & B\\textbar{}D & A\\textbar{}D & B\\textbar{}D & A\\textbar{}C\\\\\n",
       "\t6 & A\\textbar{}B & A\\textbar{}C & B\\textbar{}D & A\\textbar{}D & B\\textbar{}D & A\\textbar{}C\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | V1 &lt;chr&gt; | V2 &lt;chr&gt; | V3 &lt;chr&gt; | V4 &lt;chr&gt; | V5 &lt;chr&gt; | V6 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | A|B | A|C | B|D | A|D | B|D | A|C |\n",
       "| 2 | A|B | A|C | B|D | A|D | B|D | A|C |\n",
       "| 3 | A|B | A|C | B|D | A|D | B|D | A|C |\n",
       "| 4 | A|B | A|C | B|D | A|D | B|D | A|C |\n",
       "| 5 | A|B | A|C | B|D | A|D | B|D | A|C |\n",
       "| 6 | A|B | A|C | B|D | A|D | B|D | A|C |\n",
       "\n"
      ],
      "text/plain": [
       "  V1  V2  V3  V4  V5  V6 \n",
       "1 A|B A|C B|D A|D B|D A|C\n",
       "2 A|B A|C B|D A|D B|D A|C\n",
       "3 A|B A|C B|D A|D B|D A|C\n",
       "4 A|B A|C B|D A|D B|D A|C\n",
       "5 A|B A|C B|D A|D B|D A|C\n",
       "6 A|B A|C B|D A|D B|D A|C"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>7536</li><li>6</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 7536\n",
       "\\item 6\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 7536\n",
       "2. 6\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 7536    6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "parsed.flow %>% names\n",
    "parsed.flow[[1]] %>% head\n",
    "parsed.flow[[1]] %>% dim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "eab9ac11-37de-44ca-a91e-b2cb6c1746d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load parsed.geno\n",
    "\n",
    "parsed.geno.raw <- read.csv(file.path(args$data.dir, 'parsed_geno.csv'))\n",
    "parsed.geno <- list()\n",
    "\n",
    "for (i in 1:23) {\n",
    "    chr = paste0('chr', i)\n",
    "    parsed.geno[[i]] <- parsed.geno.raw[parsed.geno.raw$V7 == chr, ][,-7]\n",
    "}\n",
    "\n",
    "names(parsed.geno) <- chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "6cb42dde-5ac8-4801-a809-26f431076d4a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'chr1'</li><li>'chr2'</li><li>'chr3'</li><li>'chr4'</li><li>'chr5'</li><li>'chr6'</li><li>'chr7'</li><li>'chr8'</li><li>'chr9'</li><li>'chr10'</li><li>'chr11'</li><li>'chr12'</li><li>'chr13'</li><li>'chr14'</li><li>'chr15'</li><li>'chr16'</li><li>'chr17'</li><li>'chr18'</li><li>'chr19'</li><li>'chr20'</li><li>'chr21'</li><li>'chr22'</li><li>'chrX'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'chr1'\n",
       "\\item 'chr2'\n",
       "\\item 'chr3'\n",
       "\\item 'chr4'\n",
       "\\item 'chr5'\n",
       "\\item 'chr6'\n",
       "\\item 'chr7'\n",
       "\\item 'chr8'\n",
       "\\item 'chr9'\n",
       "\\item 'chr10'\n",
       "\\item 'chr11'\n",
       "\\item 'chr12'\n",
       "\\item 'chr13'\n",
       "\\item 'chr14'\n",
       "\\item 'chr15'\n",
       "\\item 'chr16'\n",
       "\\item 'chr17'\n",
       "\\item 'chr18'\n",
       "\\item 'chr19'\n",
       "\\item 'chr20'\n",
       "\\item 'chr21'\n",
       "\\item 'chr22'\n",
       "\\item 'chrX'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'chr1'\n",
       "2. 'chr2'\n",
       "3. 'chr3'\n",
       "4. 'chr4'\n",
       "5. 'chr5'\n",
       "6. 'chr6'\n",
       "7. 'chr7'\n",
       "8. 'chr8'\n",
       "9. 'chr9'\n",
       "10. 'chr10'\n",
       "11. 'chr11'\n",
       "12. 'chr12'\n",
       "13. 'chr13'\n",
       "14. 'chr14'\n",
       "15. 'chr15'\n",
       "16. 'chr16'\n",
       "17. 'chr17'\n",
       "18. 'chr18'\n",
       "19. 'chr19'\n",
       "20. 'chr20'\n",
       "21. 'chr21'\n",
       "22. 'chr22'\n",
       "23. 'chrX'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"chr1\"  \"chr2\"  \"chr3\"  \"chr4\"  \"chr5\"  \"chr6\"  \"chr7\"  \"chr8\"  \"chr9\" \n",
       "[10] \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\"\n",
       "[19] \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>V6</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>A|A</td><td>A|C</td><td>A|A  </td><td>NA|NA</td><td>A|A  </td><td>NA|NA</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>A|G</td><td>A|G</td><td>NA|NA</td><td>G|G  </td><td>G|G  </td><td>NA|NA</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>G|G</td><td>G|A</td><td>G|G  </td><td>NA|NA</td><td>G|G  </td><td>NA|NA</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>T|T</td><td>T|T</td><td>NA|NA</td><td>NA|NA</td><td>T|T  </td><td>T|T  </td></tr>\n",
       "\t<tr><th scope=row>5</th><td>T|T</td><td>T|T</td><td>NA|NA</td><td>NA|NA</td><td>T|T  </td><td>NA|NA</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>C|A</td><td>C|A</td><td>A|A  </td><td>C|A  </td><td>NA|NA</td><td>A|A  </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & V1 & V2 & V3 & V4 & V5 & V6\\\\\n",
       "  & <chr> & <chr> & <chr> & <chr> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & A\\textbar{}A & A\\textbar{}C & A\\textbar{}A   & NA\\textbar{}NA & A\\textbar{}A   & NA\\textbar{}NA\\\\\n",
       "\t2 & A\\textbar{}G & A\\textbar{}G & NA\\textbar{}NA & G\\textbar{}G   & G\\textbar{}G   & NA\\textbar{}NA\\\\\n",
       "\t3 & G\\textbar{}G & G\\textbar{}A & G\\textbar{}G   & NA\\textbar{}NA & G\\textbar{}G   & NA\\textbar{}NA\\\\\n",
       "\t4 & T\\textbar{}T & T\\textbar{}T & NA\\textbar{}NA & NA\\textbar{}NA & T\\textbar{}T   & T\\textbar{}T  \\\\\n",
       "\t5 & T\\textbar{}T & T\\textbar{}T & NA\\textbar{}NA & NA\\textbar{}NA & T\\textbar{}T   & NA\\textbar{}NA\\\\\n",
       "\t6 & C\\textbar{}A & C\\textbar{}A & A\\textbar{}A   & C\\textbar{}A   & NA\\textbar{}NA & A\\textbar{}A  \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 6\n",
       "\n",
       "| <!--/--> | V1 &lt;chr&gt; | V2 &lt;chr&gt; | V3 &lt;chr&gt; | V4 &lt;chr&gt; | V5 &lt;chr&gt; | V6 &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| 1 | A|A | A|C | A|A   | NA|NA | A|A   | NA|NA |\n",
       "| 2 | A|G | A|G | NA|NA | G|G   | G|G   | NA|NA |\n",
       "| 3 | G|G | G|A | G|G   | NA|NA | G|G   | NA|NA |\n",
       "| 4 | T|T | T|T | NA|NA | NA|NA | T|T   | T|T   |\n",
       "| 5 | T|T | T|T | NA|NA | NA|NA | T|T   | NA|NA |\n",
       "| 6 | C|A | C|A | A|A   | C|A   | NA|NA | A|A   |\n",
       "\n"
      ],
      "text/plain": [
       "  V1  V2  V3    V4    V5    V6   \n",
       "1 A|A A|C A|A   NA|NA A|A   NA|NA\n",
       "2 A|G A|G NA|NA G|G   G|G   NA|NA\n",
       "3 G|G G|A G|G   NA|NA G|G   NA|NA\n",
       "4 T|T T|T NA|NA NA|NA T|T   T|T  \n",
       "5 T|T T|T NA|NA NA|NA T|T   NA|NA\n",
       "6 C|A C|A A|A   C|A   NA|NA A|A  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>7536</li><li>6</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 7536\n",
       "\\item 6\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 7536\n",
       "2. 6\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 7536    6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "parsed.geno %>% names\n",
    "parsed.geno[[1]] %>% head\n",
    "parsed.geno[[1]] %>% dim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "c9466a39-e1b6-49a8-8a67-10e2a7b78aaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load is.corrected\n",
    "\n",
    "is.corrected.raw <- read.csv(file.path(args$data.dir, 'is_corrected.csv'))\n",
    "is.corrected <- list()\n",
    "\n",
    "for (i in 1:23) {\n",
    "    chr = paste0('chr', i)\n",
    "    is.corrected[[i]] <- is.corrected.raw[is.corrected.raw$V13 == chr, ][,-13]\n",
    "}\n",
    "\n",
    "names(is.corrected) <- chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "86205f6e-2011-4252-b3c5-ae2e3b3f498d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'chr1'</li><li>'chr2'</li><li>'chr3'</li><li>'chr4'</li><li>'chr5'</li><li>'chr6'</li><li>'chr7'</li><li>'chr8'</li><li>'chr9'</li><li>'chr10'</li><li>'chr11'</li><li>'chr12'</li><li>'chr13'</li><li>'chr14'</li><li>'chr15'</li><li>'chr16'</li><li>'chr17'</li><li>'chr18'</li><li>'chr19'</li><li>'chr20'</li><li>'chr21'</li><li>'chr22'</li><li>'chrX'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'chr1'\n",
       "\\item 'chr2'\n",
       "\\item 'chr3'\n",
       "\\item 'chr4'\n",
       "\\item 'chr5'\n",
       "\\item 'chr6'\n",
       "\\item 'chr7'\n",
       "\\item 'chr8'\n",
       "\\item 'chr9'\n",
       "\\item 'chr10'\n",
       "\\item 'chr11'\n",
       "\\item 'chr12'\n",
       "\\item 'chr13'\n",
       "\\item 'chr14'\n",
       "\\item 'chr15'\n",
       "\\item 'chr16'\n",
       "\\item 'chr17'\n",
       "\\item 'chr18'\n",
       "\\item 'chr19'\n",
       "\\item 'chr20'\n",
       "\\item 'chr21'\n",
       "\\item 'chr22'\n",
       "\\item 'chrX'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'chr1'\n",
       "2. 'chr2'\n",
       "3. 'chr3'\n",
       "4. 'chr4'\n",
       "5. 'chr5'\n",
       "6. 'chr6'\n",
       "7. 'chr7'\n",
       "8. 'chr8'\n",
       "9. 'chr9'\n",
       "10. 'chr10'\n",
       "11. 'chr11'\n",
       "12. 'chr12'\n",
       "13. 'chr13'\n",
       "14. 'chr14'\n",
       "15. 'chr15'\n",
       "16. 'chr16'\n",
       "17. 'chr17'\n",
       "18. 'chr18'\n",
       "19. 'chr19'\n",
       "20. 'chr20'\n",
       "21. 'chr21'\n",
       "22. 'chr22'\n",
       "23. 'chrX'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       " [1] \"chr1\"  \"chr2\"  \"chr3\"  \"chr4\"  \"chr5\"  \"chr6\"  \"chr7\"  \"chr8\"  \"chr9\" \n",
       "[10] \"chr10\" \"chr11\" \"chr12\" \"chr13\" \"chr14\" \"chr15\" \"chr16\" \"chr17\" \"chr18\"\n",
       "[19] \"chr19\" \"chr20\" \"chr21\" \"chr22\" \"chrX\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 12</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>V1</th><th scope=col>V2</th><th scope=col>V3</th><th scope=col>V4</th><th scope=col>V5</th><th scope=col>V6</th><th scope=col>V7</th><th scope=col>V8</th><th scope=col>V9</th><th scope=col>V10</th><th scope=col>V11</th><th scope=col>V12</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th><th scope=col>&lt;lgl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>2</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td> TRUE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>FALSE</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>FALSE</td><td>TRUE</td><td>FALSE</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 12\n",
       "\\begin{tabular}{r|llllllllllll}\n",
       "  & V1 & V2 & V3 & V4 & V5 & V6 & V7 & V8 & V9 & V10 & V11 & V12\\\\\n",
       "  & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl> & <lgl>\\\\\n",
       "\\hline\n",
       "\t1 & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE & FALSE & FALSE & FALSE & TRUE & FALSE\\\\\n",
       "\t2 & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE &  TRUE & FALSE & FALSE & FALSE & TRUE & FALSE\\\\\n",
       "\t3 & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & TRUE & FALSE\\\\\n",
       "\t4 & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & TRUE & FALSE\\\\\n",
       "\t5 & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & TRUE & FALSE\\\\\n",
       "\t6 & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & FALSE & TRUE & FALSE\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 12\n",
       "\n",
       "| <!--/--> | V1 &lt;lgl&gt; | V2 &lt;lgl&gt; | V3 &lt;lgl&gt; | V4 &lt;lgl&gt; | V5 &lt;lgl&gt; | V6 &lt;lgl&gt; | V7 &lt;lgl&gt; | V8 &lt;lgl&gt; | V9 &lt;lgl&gt; | V10 &lt;lgl&gt; | V11 &lt;lgl&gt; | V12 &lt;lgl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|---|---|---|\n",
       "| 1 | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE | FALSE | FALSE | FALSE | TRUE | FALSE |\n",
       "| 2 | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE |  TRUE | FALSE | FALSE | FALSE | TRUE | FALSE |\n",
       "| 3 | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | TRUE | FALSE |\n",
       "| 4 | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | TRUE | FALSE |\n",
       "| 5 | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | TRUE | FALSE |\n",
       "| 6 | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | FALSE | TRUE | FALSE |\n",
       "\n"
      ],
      "text/plain": [
       "  V1    V2    V3    V4    V5    V6    V7    V8    V9    V10   V11  V12  \n",
       "1 FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE TRUE FALSE\n",
       "2 FALSE FALSE FALSE FALSE FALSE FALSE  TRUE FALSE FALSE FALSE TRUE FALSE\n",
       "3 FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE TRUE FALSE\n",
       "4 FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE TRUE FALSE\n",
       "5 FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE TRUE FALSE\n",
       "6 FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE TRUE FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>7536</li><li>12</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 7536\n",
       "\\item 12\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 7536\n",
       "2. 12\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 7536   12"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "is.corrected %>% names\n",
    "is.corrected[[1]] %>% head\n",
    "is.corrected[[1]] %>% dim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "8217e68a-25d2-4675-a6b4-f8921abc96c0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load map.list\n",
    "\n",
    "map.list.raw <- read.csv(file.path(args$data.dir, 'map_list.csv'))\n",
    "map.list <- list()\n",
    "\n",
    "for (i in 1:23) {\n",
    "    chr = paste0('chr', i)\n",
    "    map.list[[i]] <- map.list.raw[map.list.raw$'chr' == chr, ][,-4]\n",
    "}\n",
    "\n",
    "names(map.list) <- chrs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "52145e53-f78b-48c3-80db-47804c29ffd7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 3</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>id</th><th scope=col>pos</th><th scope=col>pos.out</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>1</th><td>id16</td><td> 84244</td><td>84,244 </td></tr>\n",
       "\t<tr><th scope=row>2</th><td>id26</td><td>261753</td><td>261,753</td></tr>\n",
       "\t<tr><th scope=row>3</th><td>id50</td><td>665098</td><td>665,098</td></tr>\n",
       "\t<tr><th scope=row>4</th><td>id78</td><td>852019</td><td>852,019</td></tr>\n",
       "\t<tr><th scope=row>5</th><td>id79</td><td>852464</td><td>852,464</td></tr>\n",
       "\t<tr><th scope=row>6</th><td>id89</td><td>877937</td><td>877,937</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 3\n",
       "\\begin{tabular}{r|lll}\n",
       "  & id & pos & pos.out\\\\\n",
       "  & <chr> & <int> & <chr>\\\\\n",
       "\\hline\n",
       "\t1 & id16 &  84244 & 84,244 \\\\\n",
       "\t2 & id26 & 261753 & 261,753\\\\\n",
       "\t3 & id50 & 665098 & 665,098\\\\\n",
       "\t4 & id78 & 852019 & 852,019\\\\\n",
       "\t5 & id79 & 852464 & 852,464\\\\\n",
       "\t6 & id89 & 877937 & 877,937\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 3\n",
       "\n",
       "| <!--/--> | id &lt;chr&gt; | pos &lt;int&gt; | pos.out &lt;chr&gt; |\n",
       "|---|---|---|---|\n",
       "| 1 | id16 |  84244 | 84,244  |\n",
       "| 2 | id26 | 261753 | 261,753 |\n",
       "| 3 | id50 | 665098 | 665,098 |\n",
       "| 4 | id78 | 852019 | 852,019 |\n",
       "| 5 | id79 | 852464 | 852,464 |\n",
       "| 6 | id89 | 877937 | 877,937 |\n",
       "\n"
      ],
      "text/plain": [
       "  id   pos    pos.out\n",
       "1 id16  84244 84,244 \n",
       "2 id26 261753 261,753\n",
       "3 id50 665098 665,098\n",
       "4 id78 852019 852,019\n",
       "5 id79 852464 852,464\n",
       "6 id89 877937 877,937"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>7536</li><li>3</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 7536\n",
       "\\item 3\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 7536\n",
       "2. 3\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 7536    3"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "map.list[[1]] %>% head\n",
    "map.list[[1]] %>% dim"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "58a61a1a-6f3b-4e18-a677-22484fc23386",
   "metadata": {},
   "source": [
    "# Make a plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "2a33b31e-c0f1-4d6a-aa7a-ddc6c673ba62",
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = brewer.pal(brewer.pal.info[args$color.palette,]$maxcolors, args$color.palette)\n",
    "\n",
    "letters <- unique(unlist(strsplit(unique(unlist(parsed.flow)), '')))\n",
    "letters <- letters[!(letters %in% c('|', 'X'))]\n",
    "letter.colors <- c(colors[1:length(letters)], 'white')\n",
    "names(letter.colors) <- c(letters, 'X')\n",
    "\n",
    "fig <- get.haplo.profiles()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "53008cba-3bbe-457a-a2ad-57aa2d4e3c0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "haplotype_demo <- do.subplot(fig, ncol = 1, panning = .015, margin = .007, overridehovermode = 'X unified')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fac63e71-6f77-4fbc-84d8-580a445cbda5",
   "metadata": {},
   "source": [
    "# Save the plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d1acb32e-0aa8-4298-96b3-4c9bf1096e6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "saveWidget(haplotype_demo, file.path(args$out.dir, \"haplotype_demo.html\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b0d1b55-bec4-48c7-bf7e-8152a3e9b093",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
